cibuild.sh

```diff
@@ -0,0 +1,171 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+WD=$PWD
+nuttx=$WD/../nuttx
+apps=$WD/../apps
+tools=$WD/../tools
+prebuilt=$WD/../prebuilt
+
+# genromfs
+install=("if [ ! -d \"$prebuilt/genromfs/usr/bin\" ]; then \
+            if [ ! -d "$tools" ]; then \
+              git clone https://github.com/nuttx/tools.git $tools; \
+            fi; \
+            mkdir -p $prebuilt; cd $tools; \
+            tar zxf genromfs-0.5.2.tar.gz -C $prebuilt; \
+            cd $prebuilt/genromfs-0.5.2; \
+            make install PREFIX=$prebuilt/genromfs; \
+            cd $prebuilt; rm -rf genromfs-0.5.2; \
+          fi; \
+          genromfs -h; \
+          ")
+path=("$prebuilt/genromfs/usr/bin")
+
+# kconfig-frontends
+install+=("if [ ! -d \"$prebuilt/kconfig-frontends/bin\" ]; then \
+             cd $tools/kconfig-frontends; \
+             ./configure --prefix=$prebuilt/kconfig-frontends --enable-mconf --disable-gconf --disable-qconf --enable-static; \
+             make install; \
+             cd $tools; git clean -xfd; \
+           fi; \
+           ")
+path+=("$prebuilt/kconfig-frontends/bin")
+
+# arm gcc toolchain
+install+=("if [ ! -d \"$prebuilt/gcc-arm-none-eabi/bin\" ]; then \
+             cd $prebuilt; \
+             wget https://developer.arm.com/-/media/Files/downloads/gnu-rm/9-2019q4/RC2.1/gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2; \
+             tar jxf gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2; \
+             mv gcc-arm-none-eabi-9-2019-q4-major gcc-arm-none-eabi; \
+             rm gcc-arm-none-eabi-9-2019-q4-major-x86_64-linux.tar.bz2; \
+           fi; \
+           arm-none-eabi-gcc --version; \
+           ")
+path+=("$prebuilt/gcc-arm-none-eabi/bin")
+
+# mips gcc toolchain
+install+=("if [ ! -d \"$prebuilt/pinguino-compilers/linux64/p32/bin\" ]; then \
+             cd $prebuilt; \
+             git clone https://github.com/PinguinoIDE/pinguino-compilers; \
+           fi; \
+           p32-gcc --version; \
+           ")
+path+=("$prebuilt/pinguino-compilers/linux64/p32/bin")
+
+# risc-v gcc toolchain
+install+=("if [ ! -d \"$prebuilt/riscv64-unknown-elf-gcc/bin\" ]; then \
+             cd $prebuilt; \
+             wget https://static.dev.sifive.com/dev-tools/riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+             tar zxf riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+             mv riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14 riscv64-unknown-elf-gcc; \
+             rm riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+           fi; \
+           riscv64-unknown-elf-gcc --version; \
+           ")
+path+=("$prebuilt/riscv64-unknown-elf-gcc/bin")
+
+# ccache
```

Comment:
I think having the option to enable/disable ccache would be useful: ccache is used to cache compilations, but in CI is better to compile from scratch, since usually repeatability is desired despite of the speed. Or is ccache being used because the script `testbuild.sh` tries to compile the same files many times due to different configurations? If the caching is not being used disabling it would be better as the cache miss seems to have a significant performance cost (see https://ccache.dev/performance.html), and only enable it for the case a developer want to test the script locally before sending to upstream.

Reply:
I'm doing check/full build in local machine with ccache enable/disable to see the check/full build results. Still need some time.I think it would be fine to provide ccache enable/disable option.

Reply:
ok

Reply:
Add ./cibuild.sh -c option to enable ccache now. The tests in my local machine, it shows ccache would speedup 20% ~30% compile time in general, but it also failed in some case. 
Any way, we could enalbe/disable in Apache CI to test and verify later. Then decide enalbe or disable it defaultly.

sim/mips/riscv/x86 configs
============================================================
        ccache           noccache
real	12m52.185s 	15m24.466s
real	15m48.828s     15m20.406s
real   12m1.289s      15m14.861s
real   11m56.963s     15m16.283s

arm configs
============================================================
         ccache:       noccache
real	137m39.394s    182m37.709s
real	135m40.322s    184m59.780s
real	179m58.482s